### ISSUE TITLE
Can we enable SO_REUSEADDR?

### ISSUE THREAD

**Issue by NielsZeilemaker** [libswift/libswift#42]
We're having some problems with Tribler and address already in use, because they are in the time_wait state. Can we enable SO_REUSEADDR in swift?
Its already in the code, but commented out.

http://www.unixguide.net/network/socketfaq/4.5.shtml
https://github.com/libswift/libswift/blob/242d21b05859fc7aa72d2600a103575e58bce28d/channel.cpp#L266

**Comment by LipuFei** [comment]
@NielsZeilemaker No problem. I'll have a look.

**Comment by LipuFei** [comment]
@NielsZeilemaker I found this:
http://stackoverflow.com/questions/14388706/socket-options-so-reuseaddr-and-so-reuseport-how-do-they-differ-do-they-mean-t

I think the problem is that in tribler, we create a TCP socket to try to bind to an address. However, when we close the socket, the kernel does not close it immediately until the send buffer is empty or the LINGER_TIME timeout is reached. So I think, we do it in tribler instead by setting a random port TCP socket to have 0 LENGER_TIME, so that the port should be closed immediately after we call close, and then swift can bind to the random port without setting the `SO_REUSEADDR` flag.

We can first have a try with tribler to see if it works. Because as the stackoverflow post suggests, "binding a socket to exactly the same address and port as dying socket in `TIME_WAIT` state can have unexpected, and usually undesired, side effects in case the other socket is still 'at work' but that is beyond the scope of this answer and fortunately those side effects are rather rare in practice." So there might be some rare issues with `SO_REUSEADDR`.

**Comment by synctext** [comment]
Removing the comments from that line sounds like a 1-minute solution to our problem.
I hope that works; we need to get RC1 going...

**Comment by LipuFei** [comment]
@NielsZeilemaker @badrock I don't think we can use the `SO_REUSEADDR` flag here because they are UDP sockets, and `SO_REUSEADDR` only works for TCP sockets. I think we can close this issue.

**Comment by egbertbouman** [comment]
@LipuFei I think it's only the linger time the does not apply to UDP,
SO_REUSEADDR does.
